pkg/apis/messaging/v1alpha1/choice_types.go

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2019 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	eventingduckv1alpha1 "github.com/knative/eventing/pkg/apis/duck/v1alpha1"
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/apis"
+	duckv1alpha1 "knative.dev/pkg/apis/duck/v1alpha1"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+	"knative.dev/pkg/webhook"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// Choice defines conditional branches that will be wired in
+// series through Channels and Subscriptions.
+type Choice struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the Choice.
+	Spec ChoiceSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the Choice. This data may be out of
+	// date.
+	// +optional
+	Status ChoiceStatus `json:"status,omitempty"`
+}
+
+// Check that Choice can be validated, can be defaulted, and has immutable fields.
+var _ apis.Validatable = (*Choice)(nil)
+var _ apis.Defaultable = (*Choice)(nil)
+
+// TODO: make appropriate fields immutable.
+//var _ apis.Immutable = (*Choice)(nil)
+var _ runtime.Object = (*Choice)(nil)
+var _ webhook.GenericCRD = (*Choice)(nil)
+
+type ChoiceSpec struct {
+	// Cases is the list of Filter/Subscribers pairs. Filters are evaluated in the order
+	// provided, until one pass (returns true)
+	Cases []ChoiceCase `json:"cases"`
+
+	// ChannelTemplate specifies which Channel CRD to use
```

Comment:
Thanks for changing this, could you also add a comment here about the default behaviour since it's an API file.